taichi/analysis/same_statements.cpp

```diff
@@ -191,22 +214,37 @@ class IRNodeComparator : public IRVisitor {
     }
   }
 
-  static bool run(IRNode *root1, IRNode *root2) {
-    IRNodeComparator comparator(root2);
+  static bool run(IRNode *root1,
+                  IRNode *root2,
+                  std::optional<std::unordered_map<int, int>> id_map) {
+    IRNodeComparator comparator(root2, std::move(id_map));
     root1->accept(&comparator);
     return comparator.same;
   }
 };
 
 namespace irpass::analysis {
-bool same_statements(IRNode *root1, IRNode *root2) {
+bool same_statements(IRNode *root1,
+                     IRNode *root2,
+                     std::optional<std::unordered_map<int, int>> id_map) {
+  // id_map is an id map from root1 to root2.
```

Comment:
Here are we assuming `root1` and `root2` don't contain the other?

Reply:
No. We should return `false` if one of the `root`s contains the other.